posts/2023-05-17-advanced-kotlin.adoc

```diff
@@ -0,0 +1,326 @@
+= Embrace Kotlin: Where to Get Started for Scala Developers
+AnastasiiaL
+v1.0, 2023-03-31
+:title: Exploring Advanced Kotlin Features: A Deep Dive for Scala Developers
+:imagesdir: ../media/2023-03-31-embrace-kotlin
+:lang: en
+:tags: [scala, kotlin]
+
+Recently I shared my experience with trying out Kotlin coming from a Scala background. I talked about the syntax and some concept similarities.
+In this blog post, I want to focus on Kotlin's more advanced features and explore how Scala developers can leverage their existing knowledge to make the most out of Kotlin.
+
+== Kotlin's Type System: A Closer Look for Scala Developers
+
+When first diving into Kotlin from a Scala background, one of the most noticeable differences is the simplicity of Kotlin's type system.
+While it's not as intricate as Scala's, it's robust and expressive, offering everything you need for most projects.
+In this section, we will delve deeper into Kotlin's type system, variance annotations, and type projections, and draw comparisons with Scala's type system.
+
+Just like Scala, Kotlin is a statically-typed language. All types are known at compile-time, which means the compiler can catch type-related errors early.
+Unlike Scala, Kotlin does not have a universal superclass (like Any) and lower bounds, and it handles nullability differently.
+While Scala has _Option_ and _Either_ types to encapsulate optional values and potential failure cases, Kotlin tries to eliminate _NullPointerExceptions_ by distinguishing nullable and non-nullable types at the type system level.
+Each type in Kotlin is either nullable (can hold a null value) or non-nullable (cannot hold a null value). This is indicated with a question mark suffix.
+Here's a simple example:
+
+Kotlin:
+[source,kotlin]
+----
+var nonNullString: String = "Hello, World!"  // Non-Nullable String
+var nullString: String? = null  // Nullable String
+nonNullString = null // This line will cause a compile error
+----
+
+==== Variance Annotations
+Variance annotations are Kotlin's solution to handling subtype relationships for generic types.
+If you're familiar with Scala's + and - variance annotations, Kotlin's _out_ (covariant) and _in_ (contravariant) keywords will feel quite familiar.
+
+However, Kotlin imposes stricter rules on variance to ensure type safety. In Kotlin, a type parameter marked as _out_ is only available in output positions (e.g., as the return type of a function) but not in input positions (e.g., as function parameters).
+Similarly, an _in_ parameter is only allowed in input positions. This safety measure helps prevent potential runtime _ClassCastException_ errors.
```

Comment:
`at` could reduce the use of `in` - if reading out load one's tongue might trip a little, also there's a minor mix of singular and plural
```suggestion
Similarly, an _in_ parameter is only allowed at an input position. This safety measure helps prevent potential runtime _ClassCastException_ errors.
```

Or 

```suggestion
Similarly, _in_ parameters are only allowed at input positions. This safety measure helps prevent potential runtime _ClassCastException_ errors.